File: Java/Optimization/greywolf.java
import java.util.Random;

class Particle {
    double[] position;
    double[] velocity;
    double[] localBestPosition;
    double localBestFitness;
    double fitness;

    Particle(double[][] bounds, int numVariables) {
        position = new double[numVariables];
        velocity = new double[numVariables];
        localBestPosition = new double[numVariables];
        localBestFitness = Double.POSITIVE_INFINITY;
        fitness = Double.POSITIVE_INFINITY;

        Random random = new Random();
        for (int i = 0; i < numVariables; i++) {
            position[i] = bounds[i][0] + random.nextDouble() * (bounds[i][1] - bounds[i][0]);
            velocity[i] = -1 + random.nextDouble() * 2;
            localBestPosition[i] = position[i];
        }
    }

    void evaluate(double[] globalBestPosition, ObjectiveFunction objectiveFunction) {
        fitness = objectiveFunction.evaluate(position);

        if (fitness < localBestFitness) {
            localBestFitness = fitness;
            System.arraycopy(position, 0, localBestPosition, 0, position.length);
        }
    }

    void updateVelocity(double[] globalBestPosition, double inertiaConstant, double cognitiveConstant, double socialConstant) {
        Random random = new Random();
        for (int i = 0; i < velocity.length; i++) {
            double r1 = random.nextDouble();
            double r2 = random.nextDouble();

            double cognitiveVelocity = cognitiveConstant * r1 * (localBestPosition[i] - position[i]);
            double socialVelocity = socialConstant * r2 * (globalBestPosition[i] - position[i]);
            velocity[i] = inertiaConstant * velocity[i] + cognitiveVelocity + socialVelocity;
        }
    }

    void updatePosition(double[][] bounds) {
        for (int i = 0; i < position.length; i++) {
            position[i] += velocity[i];

            if (position[i] > bounds[i][1]) {
                position[i] = bounds[i][1];
            }
            if (position[i] < bounds[i][0]) {
                position[i] = bounds[i][0];
            }
        }
    }
}

interface ObjectiveFunction {
    double evaluate(double[] position);
}

class ParticleSwarmOptimization {
    void optimize(ObjectiveFunction objectiveFunction, double[][] bounds, int numVariables, int particleSize, int iterations, double inertiaConstant, double cognitiveConstant, double socialConstant) {
        Particle[] swarmParticles = new Particle[particleSize];
        double[] globalBestPosition = new double[numVariables];
        double globalBestFitness = Double.POSITIVE_INFINITY;

        for (int i = 0; i < particleSize; i++) {
            swarmParticles[i] = new Particle(bounds, numVariables);
        }

        for (int iter = 0; iter < iterations; iter++) {
            for (Particle particle : swarmParticles) {
                particle.evaluate(globalBestPosition, objectiveFunction);

                if (particle.localBestFitness < globalBestFitness) {
                    globalBestFitness = particle.localBestFitness;
                    System.arraycopy(particle.localBestPosition, 0, globalBestPosition, 0, particle.localBestPosition.length);
                }

                particle.updateVelocity(globalBestPosition, inertiaConstant, cognitiveConstant, socialConstant);
                particle.updatePosition(bounds);
            }
        }

        System.out.println("RESULT:");
        System.out.print("Optimal Solution: (");
        for (int i = 0; i < globalBestPosition.length; i++) {
            System.out.print(globalBestPosition[i]);
            if (i != globalBestPosition.length - 1) {
                System.out.print(", ");
            }
        }
        System.out.println(")");
        System.out.println("Objective function value: " + globalBestFitness);
    }

    public static void main(String[] args) {
        double[][] bounds = { {-3000000.0, 3000000.0}, {-3000000.0, 3000000.0} };
        int numVariables = 2;
        int particleSize = 10;
        int iterations = 200;
        double inertiaConstant = 0.8;
        double cognitiveConstant = 1;
        double socialConstant = 2;

        ObjectiveFunction objectiveFunction = (position) -> {
            double x1 = position[0];
            double x2 = position[1];
            return 2 * x1 * x2 + x2 - x1 * x1 - 2 * x2 * x2;
        };

        ParticleSwarmOptimization pso = new ParticleSwarmOptimization();
        pso.optimize(objectiveFunction, bounds, numVariables, particleSize, iterations, inertiaConstant, cognitiveConstant, socialConstant);
    }
}

File: Python/Optimization/greywolf.py
import random

# ------------------------------------------------------------------------------
def objective_function(position):
    x1 = position[0]
    x2 = position[1]

    z = 2*x1*x2 + x2 - x1*x1 - 2*x2*x2
    return z
  
bounds = [(-3000000, 3000000), (-3000000, 3000000)]  # upper and lower bounds of variables
num_variables = 2  # number of variables
optimization_mode = 1  # if minimization problem, optimization_mode = -1; if maximization problem, optimization_mode = 1
   
# PARAMETERS OF PSO
particle_size = 10  # number of particles
iterations = 200  # max number of iterations
inertia_constant = 0.8
cognitive_constant = 1
social_constant = 2


# ------------------------------------------------------------------------------
class Particle:
    def __init__(self, bounds):
        self.position = []  # particle position
        self.velocity = []  # particle velocity
        self.local_best_position = []  # best position of the particle
        self.local_best_fitness = initial_fitness
        self.fitness = initial_fitness
        
        for i in range(num_variables):
            self.position.append(random.uniform(bounds[i][0], bounds[i][1]))
            self.velocity.append(random.uniform(-1, 1))
  
    def evaluate(self, objective_function):
        self.fitness = objective_function(self.position)
        if optimization_mode == -1:
            if self.fitness < self.local_best_fitness:
                self.local_best_position = self.position
                self.local_best_fitness = self.fitness
        elif optimization_mode == 1:
            if self.fitness > self.local_best_fitness:
                self.local_best_position = self.position
                self.local_best_fitness = self.fitness
                
    def update_velocity(self, global_best_position):
        for i in range(num_variables):
            r1 = random.random()
            r2 = random.random()
  
            cognitive_velocity = cognitive_constant * r1 * (self.local_best_position[i] - self.position[i])
            social_velocity = social_constant * r2 * (global_best_position[i] - self.position[i])
            self.velocity[i] = inertia_constant * self.velocity[i] + cognitive_velocity + social_velocity
  
    def update_position(self, bounds):
        for i in range(num_variables):
            self.position[i] = self.position[i] + self.velocity[i]
  
            # check and repair to satisfy the upper bounds
            if self.position[i] > bounds[i][1]:
                self.position[i] = bounds[i][1]
            # check and repair to satisfy the lower bounds
            if self.position[i] < bounds[i][0]:
                self.position[i] = bounds[i][0]
  
class PSO:
    def __init__(self, objective_function, bounds, particle_size, iterations):
        global_best_fitness = initial_fitness
        global_best_position = []
        swarm_particles = []
        
        for _ in range(particle_size):
            swarm_particles.append(Particle(bounds))
        
        for _ in range(iterations):
            for particle in swarm_particles:
                particle.evaluate(objective_function)
                
                if optimization_mode == -1:
                    if particle.fitness < global_best_fitness:
                        global_best_position = list(particle.position)
                        global_best_fitness = particle.fitness
                elif optimization_mode == 1:
                    if particle.fitness > global_best_fitness:
                        global_best_position = list(particle.position)
                        global_best_fitness = particle.fitness
                        
                particle.update_velocity(global_best_position)
                particle.update_position(bounds)
                
        print('RESULT:')
        print('Optimal Solution:', global_best_position)
        print('Objective function value:', global_best_fitness)
        
if optimization_mode == -1:
    initial_fitness = float('inf')
elif optimization_mode == 1:
    initial_fitness = -float('inf')
    
PSO(objective_function, bounds, particle_size, iterations)
